torch/csrc/distributed/c10d/intra_node_comm.cu

```diff
@@ -698,5 +698,70 @@ int64_t getIntraNodeCommUsageCounter() {
   return usageCounter;
 }
 
+static __global__ void barrierKernel(
+    P2pState** p2pStates,
+    uint64_t mask,
+    size_t rank,
+    size_t worldSize) {
+  if (threadIdx.x < worldSize && (mask & (1ULL << threadIdx.x))) {
+    auto targetRank = threadIdx.x;
+    releaseSignal(&p2pStates[targetRank]->signals0[0][rank]);
+    acquireSignal(&p2pStates[rank]->signals0[0][targetRank]);
+  }
+}
+
+void IntraNodeComm::barrier(c10::optional<std::vector<int64_t>> ranks) {
+  if (!ranks.has_value()) {
+    ranks = std::vector<int64_t>(worldSize_);
+    std::iota(ranks->begin(), ranks->end(), 0);
```

Comment:
hmmm would `IntraNodeComm` always have `0 - world_size` ranks?  and those ranks are always "local ranks" instead of global rank?

For a 2-D device mesh topology, the intra node ranks are not always from 0, i.e. one intra node comm rank can be `[0, 1, 2, 3]`, another intra node comm rank can be `[4, 5, 6, 7]`

Reply:
You brought up an interesting point. I double checked the relevant semantics:
- From the perspective of a process group, the ranks are always within the context of itself
- From the perspective of `torch.distributed` APIs, the ranks are global ranks

For example, let's we have a subgroup formed by global rank `[4, 5, 6, 7]` and we'd like to broadcast from rank 4:
- With `group.broadcast`, you need to specify the source rank as 0
- With `dist.broadcast`, you need to specify the source rank as 4
- There is a translation layer in [distributed_c10d.py](https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L2114)

It's certainly a bit confusing, but it's something we'll want to respect. For the purpose of this PR, we should use group-local ranks.